python/ray/data/_internal/arrow_ops/transform_pyarrow.py

```diff
@@ -155,174 +170,248 @@ def take_table(
     return table
 
 
-def unify_schemas(
-    schemas: List["pyarrow.Schema"], *, promote_types: bool = False
-) -> "pyarrow.Schema":
-    """Version of `pyarrow.unify_schemas()` which also handles checks for
-    variable-shaped tensors in the given schemas.
+def _find_diverging_fields(schemas: List["pyarrow.Schema"]) -> Dict[str, FieldInfo]:
```

Comment:
Discussed offline:

 - We should demote this path from the hot-path to only be used as a fallback to handle our tensor types
 - By default we'd be relying on PA native `unify_schemas`
 - As such this could be substantially simplified to only search for tensor types that need to be reconciled

Reply:
I don't think current code is doing what's described here ^